First Draft and Presentation Version

CalciumKing opened the pull request:
Created a first draft of the common stock inventory management project. Some bugs remain in the project but will be fixed soon. This is the version of the project that was displayed at the zoom meeting. A core part of the project that is missing is "gs_credentials.json" which contains a private API key that is necessary for running the application. Some parts of the application are outdated to the beginning of the project such as some documentation and README.md, they will be updated in future pull requests.

## Description
Everything from the beginning of the app to the first draft presentation, this includes basic backend logic and a simple UI application.

## Motivation and Context
This is the first pull request of many to come where the app is improved and possibly expanded over time.

## How Has This Been Tested?
Basic bug testing by the creator has happened, there are currently some bugs that are yet to be fixed that cause the application to crash. A major bug that causes the program to crash is if the camera is being used in two places at once (e.g. zoom and the application, or the login page and the part scanning page), this will be fixed in future updates. As of the most recent push, there is a bug that breaks some functionality of tables, causing the order of items on the table to appear differently than what is in the database, which makes some functionality difficult to check but still works correctly on the backend.

## Where Has This Been Documented?
Primary documentation is this PR. Some files, classes, methods, etc. have docstrings attached but some may be outdated after change of functionality. Especially in newer methods, not everything has documentation. Documentation will be added before future PRs so everything is well documented in the future.

<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/a0c882c1-ef3b-47e1-9404-0d2db3654ea3" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/37e3f144-37a4-409d-95ec-4bd1b4a6a601" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/71d6e1e5-45dc-4313-8b92-e74b82a599e8" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/2594bfee-9c43-467f-aa8d-b1ffbb6a2d71" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/d839965d-d50f-48ce-b6ac-f4fd685407fb" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/80be9aae-f7e5-417f-9439-c16aab96274f" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/6a7ffeb9-5c45-435f-9089-56c9f99ce1a2" />

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [ ] New/changed functions and methods are covered in the test suite where possible
- [ ] Test suite passes locally
- [ ] Test suite passes on GitHub Actions
- [ ] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

CalciumKing committed:
Changed all instances of "appname" to "stock_manager".

CalciumKing committed:
Requested push

CalciumKing committed:
6/25 Tuesday Commit

- Finally got app to show up and sometimes change pages

- Having problem when pressing elements the app crashes and gives a random, unknown exit code

- Moving some files around, adding some to __main__.py, learning how things work professionally

CalciumKing committed:
6/26 Thursday Commit

- Responsive first page layout, window can be adjusted to any size above 800x500 and everything will proportionally change its size regardless of aspect ratio

- Added some details to __init__.py that align with python conventions, I will add more details toward the end of my project when everything is figured out

- Changed all instances of 'appname' from template to 'stock_manager' and added details for clarification or professionalism

- lots of changes to stock_manager.ui from qtDesigner

- Added slac_logo.jpg for app icon

- removed pydm from requirements.txt and added cv2 for qr code recognition and any camera needs

- added logger.py to manage the logging of information to app.log. Called it twice in __main__.py when starting and stopping the app

- In __main__.py, moved connections to function for smaller and more readable __init__ method. Changed _change_screens to not crash the app when moving between pages, added additional logic to style sideUI buttons for page clarification

CalciumKing committed:
6/27 Friday Commit

- Added __init__.py files in a few main folders to make the folders subpackages and work better with imports

- In __main__.py: removed App() and moved it to app.py, also optimized some code along the way

- Split up .ui data from one very large .ui file to multiple smaller .ui files: deleted stock_manager.ui and made view.ui, main.ui, and scanner.ui

- Created an abstract_controller.py to manage actions that all controllers do (sidebar button clicks and exiting)

- In view.py: removed the print statement and just made it load the needed sidebar data. Added table initialization to show all data in the Google sheet

- Added database.py to manage Google sheet, it will also have the methods to add and edit data. Gets all data successfully. Uses gs_credentials.json to log into the sheet.

- In requirements.txt: added gspread and oauth2client as required for database.py

- Added scanner.py to manage scanner.ui, it loads the sidebar data to switch between scanner.ui and view.ui, loads all page data correctly, displays webcam image feed in the application and can also scan for QR Codes. If it detects a QR Code it will store the data in app.log.

CalciumKing committed:
6/27 Friday Commit 2

- Added a plethora of random testing files that I've been using the last couple of days to git just in case something happens to my computer

CalciumKing committed:
Final 6/27 Friday Commit

Many .pyc files in GitHub desktop aren't shown in editor. Committing in case of importance.

CalciumKing committed:
Added Working .gitignore

CalciumKing committed:
Update .gitignore, removed all .pyc files from git

CalciumKing committed:
Docstrings + Error Handling

- Docstrings in all methods, functions, classes, files, etc.

- Basic error handling around things that don't raise errors ideally but could.

- Added return types to most methods and functions, even if return type is -> None.

- Made the possibility of GUI Error windows/popups easier to implement but didn't implement.

- Moved ui and assets folders outside stock_manager folder to follow conventions

- Added a README.md with plans for the future, will likely change or adjust later

CalciumKing committed:
Learning JIRA API

- Renamed test_google_sheets_integration.py to test_google_sheets_api.py

- Added test_jira_api.py to test jira api and compatibility

CalciumKing committed:
Wednesday 7/2 Commit, Huge Changes:

- added spacing in multiple files

- added new files to each folder's __init__.py

- added model folder with an __init__.py and item.py to make each item in a row an object, item should be iterable, subscriptable, and have a length count, all which access item.py's variables to be able to access its information as a list

- added intellij # TODO: for better task management between workdays and sessions

- made _version import shorter and more efficient in root __init__.py, hopefully it doesn't break versioneer

- commented out try-except in __main__.py because it was blocking errors, will uncomment at end of project

- added new entry types in app.log, added spacing between app sessions

- added basic error handling around files, going to add window popups eventually

- in app.py: added _create_all_items() to turn all google sheet items into objects for ease of access. added update_tables() to just update the view.ui table, but will be used in the future to update any other around application

- In database.py: added two methods to represent updating and getting information from a database, will be used in the future if we use JIRA

- Created export.py to control export.ui. can export all data to .csv, .tsv, and .psv. Will add pdf exportation eventually

- Created finish.py to control finish.ui. Page to show and customize when finalizing a form or information. Only used when subtracting items from a database but will also be used on item edit and remove. Can be customized to show error messages.

- In scanner.py: made _check_for_qr work better and complete backend tasks with scanned qr codes. Made some parts of the file more efficient

- In view.py: added table searching to filer table results and show searched information, added button to go to export.ui

- In view.ui and scanner.ui: added lots of ui elements and changes along with some minor tweaks or functionality changes.

- In .gitignore: added TODO.txt and any python files that start with ui_, those are testing files to see the application initialization process

- Created a .gitignore for exports folder, so I don't commit long data files for testing

CalciumKing committed:
7/2 second commit: simplifications, code readability changes

- Added spacing in multiple files

- Fixed imports of multiple files

- Added docstrings and type annotation to anything that needed it

- all controllers use app: 'App' and some imports to prevent infinite import loop

- In app.py: shortened long code into short * call

- In export.py: changed _sv_export to get delimiter dynamically instead of hardcoded value

CalciumKing committed:
7/3 Thursday Commit - Major Backend Changes

- In .gitignore: changed some things around, added app.log to the ignored files list so Im not adding hundreds of lines to my commit history, also app.log should be a blank file for users.

- In CODEOWNERS: just changed it to *CalciumKing because I am the owner and manager of everything in the project.

- In .idea: removed all the files so it doesnt show up on GitHub, but theyre still in my project files.

- Reformatted an __init__.py file and added to two of them.

- In view.py: shortened the file significanly by adding lots to abstract_controller.py

- In abstract_controller.py: created an __init__() method to do what every controller does, used that to make several files significantly shorter and use less imports. Made handle_side_ui a private method.

- Added edit.py to control edit.ui. Works entirely and should be completely finished

- Created constants.py to hold any constants that may be used around the program, currently only contains the total and excess equations to be used as lambdas.

- Moved, managed, created, and checked off TODO items around the program.

- In app.py: basic formatting and removals.

- In export.py: shortened with abstract_controller.py and fixed a string casing bug with exports.

- In item.py: added the None type possibility to all the variables because of the current table, these will be removed by end of project. Replaced total and excess equations with their constant values from constants.py.

CalciumKing committed:
Monday 7/7 Commit - Add/Remove Item pages, new JIRA Code

- In .gitignore: added some testing exceptions

- Added 'Add' and 'Remove' to __init__.py. Made another __init__.py dynamically get all constants from constants.py. In root __init__.py, it now sources all items from all folders as it was supposed to originally.

- Fixed imports around all files to accommodate for the new root import system with __all__ being used.

- Added error handling around various files that needed it.

- Added some comments and formatting around various files

- In abstract_controller.py: moved _handle_side_ui() to app.py.

- Created add.py to control add.ui, and remove.py to control remove.ui. remove.py works as intended, add.py hasn't been tested yet.

- In app.py: made code more organized by moving things to methods, moved a method here from abstract_controller.py.

- In constants.py: added and formatted to the file.

- In edit.py: added docstrings and dynamic changes.

- In .ui files: changed the side ui system to be more efficient and only needs to be changed in one place rather than every .ui file every change, also shortened files by styling changes but keeping the same layout.

- Created test_elon_jira_api.py to test jira api further.

CalciumKing committed:
7/8 Commit - Backend async changes

- Throughout project: added spacing, made heavy and computations methods async to work in background while other events happen onscreen. User shouldn't be stuck on one screen or have the app freeze, all computations should be done in background. Added @override to methods that are override, doesn't change functionality but does make logical sense.

- In .gitignore: adding random files that I use but shouldn't be committed, will remove at the end of project.

- In __main__.py: changed current system to be able to work with async methods.

- In export.ui, added a progress bar to show progress of download.

- In scanner.py: added a CameraThread class to handle the camera capturing separately from main program to prevent lag, errors, freezing, etc.

CalciumKing committed:
Wednesday 7/9 Commit - Better error handling and popups

- Around project: added popups to display error messages and sometimes actions or advice if something goes wrong. Made print and log statements better with capitalization and professionalism.

- In some files: basic formatting or simplification of code blocks

- In __main__.py: fixed the weird QTimer setup for a better window.run()

- In abstract_controller.py: made update_tables() async so it can be used with asyncio.gather() and update all the tables at the same time in the background

- In item.py: added __hash__ and __eq__ to be able to put Item objects into sets for better efficiency

- In scanner.py: fixed the bug that would make the camera frame blue.

CalciumKing committed:
Thursday 7/10 Commit 1 - Login and backend functionality changes

- Throughout project: added error handling and popup functionality.

- In controllers: each controller now has a PAGE_INDEX and a to_page() function instead of constantly changing screen index and changing numbers all the time, now everything is more organized.

- In scanner.py(): created Login class to control login.ui. Created an abstract scanner class because multiple pages use the camera and scan qr codes.

- In scanner.ui: Removed user functionality, the scanner now only scans items.

- In add.py: added better submit form handling and better messages and popups.

- In abstract_controller.py: created to_page(), PAGE_INDEX constants for each controller, and made a reference to database.py for ease of access with all the controllers.

CalciumKing committed:
Thursday 7/10 Commit 2 - Organization Changes

- In abstract_controller.py: added handle_connections() so controllers always have a place to put their connections, pass if otherwise or not implemented like in Login

- In controllers: moved connections to their own @override handle_connections() method and basic formatting, some warning fixes. Made some methods nested if they are only used in one place. Removed @override from any __init__ methods because technically they are but aren't override methods.

- In app.py: made a reference to screens just to get rid of some of the errors despite not needing a reference. Added nested methods for organization and one time calls. Made all the widget controllers have public references. Optimized some of the UI logic.

CalciumKing committed:
Friday 7/11 Commit 1 - Class/File Organization and Renaming (Scanners and Abstracts)

- Moved abstract methods to abstract_controller.py and renamed it to just abstract.py. Made CameraThread a nested class in AbstractScanner.

- Removed outdated docstrings.

- Renamed Scanner() in scanner.py to ItemScanner.

- Created file_exports.py and qr_generator.py for better organization and modularity. Moved all file export methods to file_exports.py.

- Fixed imports and added new files and classes to __all__ in __init__.py files.

CalciumKing committed:
Friday 7/11 Commit 2 - Project Hierarchy Organization

- It is convention to organize even the tests folder, so I moved each file into its own folder to either test functionality of a single feature or api testing, both should be added to in the future as usual.

- Deleted __init__.py from tests because it wasn't used and after moving everything into folders, there's no point to it anyway.

CalciumKing committed:
Requested Updated Commit

CalciumKing committed:
Merge branch 'pcdshub:master' into master

ZLLentz commented:
There's a bit too much here to do a thorough line-by-line sort of review as would be standard, and probably there's a lot of value in merging the demo code as is.

I'm going to do a general high-level review with the focus on making sure there's nothing structurally wrong with the repo.

ZLLentz commented on .github/CODEOWNERS:
Change request: GitHub is really upset with this file because your user isn't a member of the pcdshub org.

A few things:
- Is there a reason we haven't added him to the org @slactjohnson? I'm not sure what the SOP is, we didn't add any of the other interns to the org either.
- It's not correct to remove the code review groups entirely. The implication of the edit here is that you'll be reviewing all of your own code changes, including org-level repo settings, which doesn't make any sense.
- I think it makes sense to keep reviews from python-reviewers (we can discuss the expectations for this if you want) but you must restore the ownership of the .github settings folder.

slactjohnson replied:
@ZLLentz I don't really know what the SOP is either, I guess I don't have strong feelings one way or the other.

ZLLentz replied:
IMO:

- I'll add @CalciumKing as a contributor to this specific repo (write permission)
- All changes to CODEOWNERS should be reverted
- We should do some light code review, to not slow Landen down but to make sure we can continue maintaining this into the future

tangkong replied:
I'm in agreement with Zach's plan here.  In case it wasn't clear from my review comments, I don't think all my comments should be resolved before merging.  I just wanted to make sure I called out issues I saw for future us

ZLLentz commented on README.md:
Comment: I'm a fan of how you've laid out this readme :+1:

ZLLentz commented on exports/.gitignore:
Question: I hadn't seen this before (subdirectory `.gitignore`). Do you find this more convenient than continuing to add rules to the top-level `.gitignore`? I'm curious because it seems like it could be a useful pattern.

tangkong replied:
It seems more annoying to me to have a fragmented source of gitignore information, but maybe I also don't see the utility quite yet.

CalciumKing replied:
I added the subdirectory `.gitignore` because git tracks files not folders, and empty folders are automatically ignored by git. I wanted the folder to be empty but I wanted it to still exist so I added a `.gitignore` file that is responsible for ignoring all files in that subdirectory except itself, allowing git to track the folder. If I had added an exception in the root `.gitignore` to ignore all files in that folder, it would have also ignored the folder. I agree that having fragmented `.gitignore` sources would be annoying, but I figured this would be an exception. I don't plan on making any more `.gitignore` files.

ZLLentz replied:
This makes sense to me, thanks for the explanation.

I think in the past I've made an empty `.gitkeep` file to do the same, what I like about `.gitkeep` is that it's somewhat clear why the file exists (to make git keep the directory).

But this works as-is so I'm not going to dwell further.

CalciumKing replied:
I've never heard of `.gitkeep` but after doing some research that it very good to know.

ZLLentz commented on launcher.sh:
Question/Nitpick: Does this application use pydm at all? Maybe a lot of these files are just vestigial from the original template used and could be removed in a future update.

CalciumKing replied:
No, this application uses no pydm, I likely just missed a few instances from the template.

ZLLentz commented on setup.py:
```diff
@@ -9,18 +9,25 @@
+            'stock_manager=stock_manager.__main__:main'
         ]
     },
+    python_requires='>=3.8',
```
Nitpick: this python version contradicts the >=3.10 from your readme

ZLLentz reviewed (changes requested):
We should definitely merge the code without straying far from the build you showed at the zoom meeting, that's an important milestone and you've made a ton of great progress here.

Going further we should decide how much/often to do code reviews in this repo given that we'll want to keep running/maintaining this app into the future. (I didn't look very closely today, happy Friday!).

The only reason this is a "Request changes" and not an "Approve" is because the `CODEOWNERS` file is now broken. It's pretty bad for this file to be broken on `master`. I didn't have any other objections to the PR.

CalciumKing committed:
Docstrings, Basic Formatting Changes, TODO Updates

- Added docstrings to everything throughout entire project.

- Reformatted some files or changed basic functionality, but everything still works.

- Updated or deleted TODOs around project

CalciumKing committed:
Enum Functionality, No Need For PAGE_NAMES in constants.py

- Removed the use of PAGE_NAMES from constants.py and replaced with use of Pages enums from enums.py.

- Removed the use of PAGE_INDEX from abstract.py and also replaced with enum functionality.

- Added functionality so that Pages enums are used for window titles, file names, and page indexes. Possibility to use this same system to only display certain pages to admin members.

CalciumKing committed:
Converted Some Lists To Sets, Improved Docstrings

- Where lists were used and sets could be used instead I updated them and made everything work more efficiently.

- In app.py: shortened some code and improved docstrings.

- In database.py: converted a list to a set and modified docstrings

tangkong commented on conda-recipe/meta.yaml:
```diff
@@ -17,13 +15,19 @@ requirements:
 
 about:
-  home: https://github.com/pcdshub/pydm-application-template
+  home: https://github.com/CalciumKing/SLAC-LCLS-Stock-Management.git
```
this should probably be the pcdshub repo now right?

tangkong commented on stock_manager/controllers/abstract.py:
```diff
@@ -0,0 +1,233 @@
+	A metaclass combining PyQts QWidget metaclass and Python's ABCMeta.
+	
+	This enables the AbstractController to inherit from both PyQt widgets and Python abstract base classes,
+	resolving metaclass conflicts that would otherwise occur with multiple inheritance.
```
Could you explain what errors you were running into here w.r.t. multiple inheritance?  You have a more complicated base widget class than we normally use, and I'm interested in what this is doing for you.

CalciumKing replied:
When I don't use `metaclass=CombinedMeta`, I get this error:
`TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases`
The idea behind the abstract classes is to only write methods once, so similar or even identical methods are not written between controller classes, and to make the process for creating new controllers easier. As seen in `view.py` for example, most of the work is being done in `abstract.py`.
While having abstract controllers may be more complicated than is needed, in my experience it makes the process for making new controllers easier.

tangkong replied:
I guess a more pointed question here is: why are we inheriting from metaclasses instead of just normal classes?  I'd expect simple QWidget subclasses to work just fine.

I'm all for minimizing code duplication, I've just found metaclasses to be a bit more deep python magic than is necessary most of the time.

CalciumKing replied:
It is very likely that there is a more efficient way of doing this, however I do not know it. I am still learning python abstract classes, as it is very different to java abstract classes that I am used to. If I find a better solution or someone here knows it, I am very open to learning and implementing it, however for the time being I am just focusing on getting the application working.

tangkong commented on stock_manager/controllers/add.py:
```diff
@@ -0,0 +1,121 @@
+			self.manufacturer,
+			self.desc
+		]
+		self.PAGE_INDEX = 3
```
Beware magic numbers here.  If we want to change this index, we have to search through the rest of the page files and manually verify that the others are still correct.

In the future I think it's worth setting up a more automatic way of assigning the page indices

tangkong commented on stock_manager/controllers/abstract.py:
```diff
@@ -0,0 +1,233 @@
+		pass
+	
+	@staticmethod
+	def filter_table(text: str, table: QTableWidget) -> None:
```
Qt has a FilterProxyModel for handling behavior like this.  It's a bit more complicated to set up, but it is a bit more robust than manually showing / hiding rows
https://doc.qt.io/qt-6/qsortfilterproxymodel.html

tangkong commented on stock_manager/app.py:
```diff
@@ -0,0 +1,189 @@
+	def __init__(self):
+		"""Initialize the main application window and setup screens."""
+		
+		from stock_manager import DBUtils, Logger, Edit, Export, Finish, Remove, ItemScanner, View, Add, Login, \
```
Throughout the project you import sub-packages in methods.  Is there a reason you don't do it all up front at the top of the file?  This is usually done if the imports are infrequent and incur a performance hit, but in our case many of these imports are in common methods.

CalciumKing replied:
When I import any of those methods at the top of the file I get "Circular Import Errors". I haven't worked with languages that have that kind of error, this was something that I found fixed the issue, however I agree that this could be done better.

tangkong replied:
Ah a classic.  Python evaluates the entire file whenever you import a module from it, so if two modules rely on each other's code you'll see this circular import error.  Typically you get around this by combining modules that do this, since the fact that they rely on each other means their code is related.

In this case it seems to be because you're using `App` for type hinting, though since you've wrapped that import in a `if TYPE_CHECKING` block I'd actually expect this to work fine at the top of the file.

CalciumKing replied:
I've been able to resolve some, but I believe that I get the issue because of the way my `__init__.py` files are set up, at least that is what my code error messages say.

tangkong commented on stock_manager/tests/api/test_elon_jira_api.py:
```diff
@@ -0,0 +1,312 @@
+import asyncio
```
(for some reason I can't comment on the file as a whole, there is nothing wrong with this line in particular)

Generally this folder is for unit tests, but it seems like you've placed a bunch of scratch work / scripts here instead.  It's good to keep these files around, but we'll have to move them eventually

CalciumKing replied:
Yes, I figured that I would delete them before the end of the project, I simply have them for testing individual features or if I need to isolate an issue for testing before bringing a solution to the main code. I figured that is what the `test` folder is for as I've seen in other projects, but I wasn't sure.

tangkong replied:
Sometimes we make a `scripts` folder for holding this like this, maybe that's a place these can live on even after you're done using them.

CalciumKing replied:
What is the difference between a `scripts` folder and `tests`? I would think that these should stay in `test` because they are testing individual features so they don't break the main program. Also aren't all the files technically "scripts", why would only these ones go in a `scripts` folder?

tangkong replied:
By tests we mean unit tests, which programmatically test our code for regressions.  These tests are gathered in continuous integration and run automatically.  We typically run these on github actions, and also use them for integration tests with other applications

This is a little different from scripts (or specifically the scripts here), which are typically run manually

tangkong commented on stock_manager/utils/constants.py:
```diff
@@ -0,0 +1,22 @@
+from typing import Callable
+
+TOTAL_EQUATION: Callable[[int, int], int] = \
```
If you're reusing a function, you should just define it.  Leaving it as a lambda prevents tracebacks from providing useful information, and it's not actually that much more typing at this point.

tangkong commented on stock_manager/utils/logger.py:
```diff
@@ -0,0 +1,51 @@
+import logging as log
+
+
+class Logger:
```
it looks like we could have just used the built-in logger here.  We usually invoke it in a way similar to below, without the extra wrapper.  Normally we set up the logging config in the app entrypoint, so it applies to the entire app once it starts

```python
# initialize a logger for the submodule, or name it 
logger = logging.getLogger(__name__)

# call logger directly in methods or class methods, etc
logger.warn("warning message")
```

Nothing broken about this, just seems a bit extraneous

tangkong commented on stock_manager/utils/enums.py:
```diff
@@ -0,0 +1,31 @@
+from enum import Enum
+
+
+class PageNames(Enum):
```
I'm happy to see thorough use of enums, it's very helpful for readability 👍

tangkong reviewed (commented):
I'm also giving this a skim, I think on the whole this looks really good, and as Zach said we should merge and fix some things in post.  

I'm going to list some general issues here, most related to getting this working in our stack
* We use qt5 at LCLS.  It'd be even better here if we used the `qtpy` package, which is a translation layer that makes our code agnostic to the precise qt binding.  I've seen the qt5->qt6 migratio effort be quite painful, so it's probably best to make this change early.
* This needs a general style pass.  We abide by pep8 standards broadly
  * tabs -> spaces (4 spaces)

I realize I've left a lot of suggestions / criticisms, but I'd consider most of them nitpicks.  This looks really good on the whole, and it's refreshing to see someone build a qt app that hasn't been exposed to our conventions.  

Nice work overall!

CalciumKing committed:
Google Sheets API, Backend Logic

- Made functioning Google Sheets API Logic for adding, updating, and removing items from the database in database.py.

- Used DatabaseUpdateType enum in files that use update_database()

- Gave DBUtils a reference to app

- In database.py: along with above changes, made API get correct sheet by using strings rather than .sheet1.

- In test_google_sheets_api.py: played around with some features, continued testing in that file.

CalciumKing committed:
Fixed issues with CODEOWNERS

CalciumKing committed:
Fixed setup.py python edition issue

ZLLentz reviewed (dismissed):
The CODEOWNERS is valid now according to github and has python-sme and software-admin, so I'm revising my review to "approve" :+1:

CalciumKing committed:
Fixed issues with meta.yaml, changed home information to pcdshub/stock-management.git

CalciumKing committed:
Fixed issue with constants.py lambdas

- Turned constants.py lambdas into functions and changed references throughout project.

CalciumKing committed:
Added status enum to table columns

- Admin can now easily see what needs to be reordered instead of viewing several numbers.

- Numbers correlated to status are currently 1, 0, 0>, which can all be updated later if needed.

CalciumKing committed:
Made a success indicator on item create, Async data fetching

- In add.py: made field checking a nested function for organization and clarity, added a check to see if that item already exists in the database, organized some code better, and made a finish screen like in scanner.py as per Jasmine's request.

ZLLentz commented:
@CalciumKing how do you prefer to proceed: do you want to make additional edits based on feedback, or do you prefer to merge now and handle anything else as part of follow-up work? I think there were no other blocking comments.

CalciumKing committed:
Convention Fixes, Organization, Import Fixes

- Throughout project: changing "" strings to '', shortening imports or changing what sub-package imports won't throw errors, basic formatting

- In edit.py: deleted _get_selected_item() because it isn't used anywhere.

- In export.py: changed multi case statement from , to |. Although it worked, it is better to use | apparently.

CalciumKing commented:
In my most recent push I have made edits based on feedback. I feel the edits I made resolved some of the feedback I was given, however some unresolved features will be fixed in future PRs. As of now I think we can safely merge and I can continue to work on feedback and developing the app further.

ZLLentz commented:
Ok, I will re-approve and click "merge"
